### DISCUSSION TITLE
Shuffling with multiple constraints (+ data output)

### DISCUSSION THREAD

**Discussion by ghost** [jspsych/jsPsych#762] [Q&A] [answered]
Hi Josh, I have two follow-up questions to the [shuffling with constraint](https://github.com/jspsych/jsPsych/discussions/706) and [conditional trial after every Nth item in timeline](https://github.com/jspsych/jsPsych/discussions/716) questions you previously answered (thanks again for those answers, big help!) 

The experiment as it is coded now is a dual task experiment with a 1I-4AFC perception task and a free recall task. There are 96 stimuli in total (which are shuffled with no immediate repeats) and after every 8th trial there is a free recall where participants have to write down as many names of the previous 8 stimuli as they can. Our code (based on your suggestions from my previous two questions) now looks like this:

```js
timeline.push({
    timeline: [
    {
        type: 'html-keyboard-response',
        stimulus: "<div style='font-size:32px'>+</div>",
        choices: jsPsych.NO_KEYS,
        trial_duration: 1000
      },
      { 
        type: 'audio-keyboard-response',
        stimulus: jsPsych.timelineVariable('audio'),
        choices: jsPsych.NO_KEYS,
        trial_ends_after_audio: true
      },
      { 
        type: 'html-button-response',
        stimulus: "<p>Which question did the speaker answer?<br></p>",
        choices: [jsPsych.timelineVariable('questionSF'), 
                  jsPsych.timelineVariable('questionVF'), 
                  jsPsych.timelineVariable('questionOF'),
                  jsPsych.timelineVariable('questionBF')],
        button_html: '<button class="jspsych-btn-block">%choice%</button>', 
        margin_vertical: '12px',
        data: jsPsych.timelineVariable('data'),
        on_finish: function(data){
          data.trial = jsPsych.data.get().filter({loop: 'trial', task: 'response'}).count();
          if(data.button_pressed == '0'){
            data.response = 'SF';
          } else if (data.button_pressed == '1'){
            data.response = 'VF';
          } else if (data.button_pressed == '2'){
            data.response = 'OF';
          } else {
            data.response = 'BF';
          }
          if(data.button_pressed == data.button){
            data.response_correct = true;
          } else {
            data.response_correct = false;
          }
        }
      },
      {
        timeline: [{
          type: 'survey-text',
          questions: [
            { prompt: "<p>Please recall as many names of the previous 8 sentences as you can.<br>" + "You can type in the names in any order you prefer (separated by a space):</p>", placeholder: "e.g. Lisa Tom", name: 'recall', required: true }
          ],
          button_label: 'Continue',
          data: {loop: 'trial', task: 'recall'},
          on_finish: function(data){
            data.recall = JSON.parse(data.responses).recall.split(" ");
          }
        }],
        conditional_function: function(){
            if(jsPsych.data.get().filter({loop: 'trial', task: 'response'}).count() % 8){
              return false;
            } else {
              return true;
            }
        }
      }
    ],
    timeline_variables: [
        { audio: 'audio/T1-SF-S1F.wav', data: {button: '0', loop: 'trial', task: 'response', sentence: 'T1', focus: 'SF', speaker: 'S1F', name: 'John'}, questionSF: '"<b>Who</b> kicked the ball?"', questionVF: '"What did John <b>do</b> with the ball?"', questionOF: '"<b>What</b> did John kick?"', questionBF: '"What happened"?' },
        { audio: 'audio/T1-VF-S2F.wav', data: {button: '1', loop: 'trial', task: 'response', sentence: 'T1', focus: 'VF', speaker: 'S2F', name: 'John'}, questionSF: '"<b>Who</b> kicked the ball?"', questionVF: '"What did John <b>do</b> with the ball?"', questionOF: '"<b>What</b> did John kick?"', questionBF: '"What happened"?' },
        { audio: 'audio/T1-OF-S4M.wav', data: {button: '2', loop: 'trial', task: 'response', sentence: 'T1', focus: 'OF', speaker: 'S4M', name: 'John'}, questionSF: '"<b>Who</b> kicked the ball?"', questionVF: '"What did John <b>do</b> with the ball?"', questionOF: '"<b>What</b> did John kick?"', questionBF: '"What happened"?' },
        { audio: 'audio/T1-BF-S3M.wav', data: {button: '3', loop: 'trial', task: 'response', sentence: 'T1', focus: 'BF', speaker: 'S3M', name: 'John'}, questionSF: '"<b>Who</b> kicked the ball?"', questionVF: '"What did John <b>do</b> with the ball?"', questionOF: '"<b>What</b> did John kick?"', questionBF: '"What happened"?' },

        { ... }, // other 22x4=88 stimuli removed to save space

        { audio: 'audio/T24-SF-S1F.wav', data: {button: '0', loop: 'trial', task: 'response', sentence: 'T24', focus: 'SF', speaker: 'S1F', name: 'Susan'}, questionSF: '"<b>Who</b> ate the apple?"', questionVF: '"What did Susan <b>do</b> with the apple?"', questionOF: '"<b>What</b> did Susan eat?"', questionBF: '"What happened"?' },
        { audio: 'audio/T24-VF-S4M.wav', data: {button: '1', loop: 'trial', task: 'response', sentence: 'T24', focus: 'VF', speaker: 'S4M', name: 'Susan'}, questionSF: '"<b>Who</b> ate the apple?"', questionVF: '"What did Susan <b>do</b> with the apple?"', questionOF: '"<b>What</b> did Susan eat?"', questionBF: '"What happened"?' },
        { audio: 'audio/T24-OF-S2F.wav', data: {button: '2', loop: 'trial', task: 'response', sentence: 'T24', focus: 'OF', speaker: 'S2F', name: 'Susan'}, questionSF: '"<b>Who</b> ate the apple?"', questionVF: '"What did Susan <b>do</b> with the apple?"', questionOF: '"<b>What</b> did Susan eat?"', questionBF: '"What happened"?' },
        { audio: 'audio/T24-BF-S3M.wav', data: {button: '3', loop: 'trial', task: 'response', sentence: 'T24', focus: 'BF', speaker: 'S3M', name: 'Susan'}, questionSF: '"<b>Who</b> ate the apple?"', questionVF: '"What did Susan <b>do</b> with the apple?"', questionOF: '"<b>What</b> did Susan eat?"', questionBF: '"What happened"?' },
    ],
    sample: {
      type: 'custom',
      fn: function(){
            var groups = [[0,1,2,3],[4,5,6,7],[8,9,10,11],[12,13,14,15],[16,17,18,19],[20,21,22,23],[24,25,26,27],[28,29,30,31],[32,33,34,35],[36,37,38,39],[40,41,42,43],[44,45,46,47],[48,49,50,51],[52,53,54,55],[56,57,58,59],[60,61,62,63],[64,65,66,67],[68,69,70,71],[72,73,74,75],[76,77,78,79],[80,81,82,83],[84,85,86,87],[88,89,90,91],[92,93,94,95]];
            var all_items = [];
            for(var i=0; i<groups.length; i++){ 
                    for(var j=0; j<groups[i].length; j++){
                                all_items.push({group: i, trial_num: groups[i][j]});
                    }
            }
            var order = jsPsych.randomization.shuffleNoRepeats(all_items, function(a,b){ 
                    return a.group == b.group;
            });
            return order.map(function(x){ return x.trial_num }); 
      }
    }  
  }); 
```

I have two follow-up questions, which I hope you can help me with:

**1. Shuffling with multiple constraints**
When I posted the [shuffling with constraint](https://github.com/jspsych/jsPsych/discussions/706) question, we didn't have the addition of the free recall task and the shuffle without repeats was exactly what we were looking for. But now that we've added the free recall task, we need an additional constraint. Namely, the block of 8 trials after which participants have to do the free recall task should not contain any repeat names. The group members defined in the groups array in the sample function indicate the stimuli with the same name in the sentence (e.g. first 4 stimuli have the name 'John', [...], final 4 stimuli have the name 'Susan'). For the free recall task to work we need to make sure the names are unique and thus that items from a single group array should not be presented within the same block of 8 (but otherwise still completely randomised without immediate repeats). Hope I'm explaining myself well. Would this be possible?

**2. Data output**
On a related note, I was wondering whether it would be possible to add the responses given during the free recall task to the previous 8 trials in the output file. The response are split into separate columns using `data.recall = JSON.parse(data.responses).recall.split(" ");`, but are still in their own row. Would it be possible to somehow automatically copy and paste the content of these columns to the previous 8 trials? That would making matching the response to the actual names a bit easier. This is not very urgent, but it would be nice to have jsPsych do this automatically rather than having to do it manually.

Thanks!

**Comment by ghost** [comment, marked as answer]
Any suggestions @jodeleeuw?

**Reply by ghost** [reply]
Perhaps a solution might be to somehow loop through a sampling function where the groups are shuffled and a random item from the first 8 groups is selected ...

```js
sample: {
      type: 'custom',
      fn: function(){
            var groups = [[0,1,2,3],[4,5,6,7],[8,9,10,11],[12,13,14,15],[16,17,18,19],[20,21,22,23],[24,25,26,27],[28,29,30,31],[32,33,34,35],[36,37,38,39],[40,41,42,43],[44,45,46,47],[48,49,50,51],[52,53,54,55],[56,57,58,59],[60,61,62,63],[64,65,66,67],[68,69,70,71],[72,73,74,75],[76,77,78,79],[80,81,82,83],[84,85,86,87],[88,89,90,91],[92,93,94,95]];
            var order = jsPsych.randomization.shuffle(groups); // shuffle groups
            var selected_items = [];
            for(var i=0; i<8; i++){ // iterate through first 8 groups
                selected_items.push(order[i][Math.round(Math.random())]) // pick one item at random from each group
            }
            return selected_items;
      }
    }
```

... and then have the selected items deleted from the (nested) array, reshuffle the groups, and again select one item at random from the first 8 groups, delete, reshuffle, select, etc. until the array is empty. Thus ensuring that each item in a sequence of 8 trials is selected from separate groups without fixing the group order or order of items within groups.

Anyone any ideas on how to accomplish this?

**Reply by jodeleeuw** [reply]
Hi @Marita-K-Everhardt, sorry I'm a little slow here. Trying to catch up on these open threads.

I think I need to clarify the end goal here before I can offer clear advice:

1. You've got **24** different groups, each with 4 trials.
2. You'd like to show blocks of **8** trials, with each trial being from a different group.
3. You want to present all **96** trials consecutively.

Is that correct?

**Reply by ghost** [reply]
Hi @jodeleeuw, no worries, I understand, busy times! Appreciate you taking the time to help.

To answer your questions: yes, yes, and yes :)

**Reply by jodeleeuw** [reply]
Great! I think your method is a good approach. One way to simplify it would be to randomize the order within each group first. That way you can just remove the first item when you use that group. The other thing that I think you'll need to do is make the `groups` variable global, so you can reference the same set of groups with each block.

```js
// move groups out of any function call
var groups = [[0,1,2,3],[4,5,6,7],[8,9,10,11],[12,13,14,15],[16,17,18,19],[20,21,22,23],[24,25,26,27],[28,29,30,31],[32,33,34,35],[36,37,38,39],[40,41,42,43],[44,45,46,47],[48,49,50,51],[52,53,54,55],[56,57,58,59],[60,61,62,63],[64,65,66,67],[68,69,70,71],[72,73,74,75],[76,77,78,79],[80,81,82,83],[84,85,86,87],[88,89,90,91],[92,93,94,95]];
// shuffle the internal order within each group
for(var i=0; i<groups.length; i++){
	groups[i] = jsPsych.randomization.shuffle(groups[i]);
}

sample: {
	type: 'custom',
	fn: function(){
            	var order = jsPsych.randomization.shuffle(groups); // shuffle groups
            	var selected_items = [];
		var which_group = 0;
		// use a while loop instead of a for loop because it lets us skip groups that have already used all their items.
            	while(selected_items.length < 8){
			// check that this group has items left
			if(order[0].length > 0) {
				// pick one item at random from each group; using .pop() removes the item from the group.
				selected_items.push(order[i].pop());
			}
			which_group++;
            	}
            return selected_items;
      }
}
```

**Reply by ghost** [reply]
I get a TypeError: Cannot read property 'pop' ...

<img width="556" alt="TypeError" src="https://user-images.githubusercontent.com/65547758/86338631-d7acfd00-bc52-11ea-8a66-6d0cbc1b399d.png">

**Reply by jodeleeuw** [reply]
Ah, my mistake. 

This line...
```js
selected_items.push(order[i].pop());
```

...should be
```js
selected_items.push(order[which_group].pop());
```

**Reply by ghost** [reply]
That solves the error, thanks! But now the trial loop stops after 8 trials, so not all 96 trials are presented consecutively.

**Reply by jodeleeuw** [reply]
Ah, I misunderstood. I though that you were hoping to put a recall trial in between every 8 trials. Is that not right?

**Reply by ghost** [reply]
Yes, but that is already in the loop with a `conditional_function`. The current code (see below) gives me 8 trials, then a free recall, and then the experiment finishes. So I think all that is left is adding something to the `sample` function that makes sure all 96 trials are selected, right? Or does this only work if I push the entire loop to the timeline 12 times? 

```js
var groups = [[0,1,2,3],[4,5,6,7],[8,9,10,11],[12,13,14,15],[16,17,18,19],[20,21,22,23],[24,25,26,27],[28,29,30,31],[32,33,34,35],[36,37,38,39],[40,41,42,43],[44,45,46,47],[48,49,50,51],[52,53,54,55],[56,57,58,59],[60,61,62,63],[64,65,66,67],[68,69,70,71],[72,73,74,75],[76,77,78,79],[80,81,82,83],[84,85,86,87],[88,89,90,91],[92,93,94,95]];
for(var i=0; i<groups.length; i++){
	groups[i] = jsPsych.randomization.shuffle(groups[i]);
}

timeline.push({
    timeline: [
      { ... }, // fixation cross removed to save space (available in original post above)
      { ... }, // audio stimulus removed to save space (available in original post above)
      { ... }, // button response removed to save space (available in original post above)
      { // this is the free recall with a conditional function so it occurs after every 8th trial
        timeline: [{
          type: 'survey-text',
          questions: [
            { prompt: "<p>Please recall as many names of the previous 8 sentences as you can.<br>" + "You can type in the names in any order you prefer (separated by a space):</p>", placeholder: "e.g. Lisa Tom", name: 'recall', required: true }
          ],
          button_label: 'Continue',
          data: {loop: 'trial', task: 'recall'},
          on_finish: function(data){
            data.recall = JSON.parse(data.responses).recall.split(" ");
          }
        }],
        conditional_function: function(){
            if(jsPsych.data.get().filter({loop: 'trial', task: 'response'}).count() % 8){
              return false;
            } else {
              return true;
            }
        }
      }
    ],
    timeline_variables: [ ... ], // timeline variables removed to save space (available in original post above)
    sample: {
	type: 'custom',
	fn: function(){
            	var order = jsPsych.randomization.shuffle(groups);
            	var selected_items = [];
		var which_group = 0;
            	while(selected_items.length < 8){
			if(order[0].length > 0) {
				selected_items.push(order[which_group].pop());
			}
			which_group++;
            	}
            return selected_items;
      }
  }  
});
```

**Reply by jodeleeuw** [reply]
Got it! Having a hard time keeping track of the setup, but this makes total sense. In that case I think you just need to repeat the sampling process 12 times.

```js
sample: {
	type: 'custom',
	fn: function(){
		var final_set = []; // create empty array to hold items.
		for(var i=0; i<12; i++){ // repeat 12 times
            		var order = jsPsych.randomization.shuffle(groups);
            		var selected_items = [];
			var which_group = 0;
            		while(selected_items.length < 8){
				if(order[0].length > 0) {
					selected_items.push(order[which_group].pop());
				}
				which_group++;
            		}
			final_set = final_set.concat(selected_items); // add the items to the end of the array
		}
		return final_set;
	}
}
```

**Reply by ghost** [reply]
Yeah, it's pretty complicated, sorry about that! 

This sounds like an easy solution.. but now my experiment won't load at all unfortunately.

**Reply by jodeleeuw** [reply]
No error messages in the console? Is it loading indefinitely?

**Reply by ghost** [reply]
Correct, no error messages, just loading indefinitely on a blank white page (doesn't even start the preloading progress bar).

**Reply by jodeleeuw** [reply]
I think I caught another bug. Here's a hopefully fixed version:

```js
sample: {
	type: 'custom',
	fn: function(){
		var final_set = []; // create empty array to hold items.
		for(var i=0; i<12; i++){ // repeat 12 times
            		var order = jsPsych.randomization.shuffle(groups);
            		var selected_items = [];
			var which_group = 0;
            		while(selected_items.length < 8){
				if(order[which_group].length > 0) {
					selected_items.push(order[which_group].pop());
				}
				which_group++;
            		}
			final_set = final_set.concat(selected_items); // add the items to the end of the array
		}
		return final_set;
	}
}
```

**Reply by ghost** [reply]
That gives me another TypeError: Cannot read property 'length' for the line `if(order[which_group].length > 0) {`

**Reply by jodeleeuw** [reply]
I did some testing instead of just trying to make up code on the fly. It turns out this algorithm isn't guaranteed to work because you could end up in a situation where for one of the final sets of 8 there aren't 8 unique groups left. So something fancier is needed to ensure a good distribution of the groups. What kinds of constraints do you want to put on the process here? Do you want to sample once from each group before repeating a group? Another option, if you want a random distribution, would be to repeat this random sampling process over and over until a valid solution is found. I'm not sure how long that might take, but I'd guess it will probably be fast enough to be unnoticeable.

**Reply by ghost** [reply]
There's no need to sample once from each group before repeating a group. The distribution can be random, as long as the sets of 8 are from separate groups. So trials 1–8 should come from separate groups, and trials 9–16 should come from separate groups, but it's  fine if for instance trial 8 and trial 9 come from the same group since they are not in the same set of 8.

**Reply by jodeleeuw** [reply]
In that case I think the simplest solution is just to try re-randomizing it many times until one works. This should do that. In my tests it can run this function fast enough that the user won't notice it taking time to re-randomize.

```js
sample: {
  type: 'custom',
  fn: function(){
    var success = false
    while(!success){
      success = true;

      var groups = [[0,1,2,3],[4,5,6,7],[8,9,10,11],[12,13,14,15],[16,17,18,19],[20,21,22,23],[24,25,26,27],[28,29,30,31],[32,33,34,35],[36,37,38,39],[40,41,42,43],[44,45,46,47],[48,49,50,51],[52,53,54,55],[56,57,58,59],[60,61,62,63],[64,65,66,67],[68,69,70,71],[72,73,74,75],[76,77,78,79],[80,81,82,83],[84,85,86,87],[88,89,90,91],[92,93,94,95]];
      for(var i=0; i<groups.length; i++){
        groups[i] = jsPsych.randomization.shuffle(groups[i]);
      }
      
      var final_set = []; // create empty array to hold items
      for (var i = 0; i < 12; i++) { // repeat 12 times
        var order = jsPsych.randomization.shuffle(groups);
        var selected_items = [];
        var which_group = 0;
        while (selected_items.length < 8) {
          if(which_group == order.length){
            success=false;
            break;
          }
          if (order[which_group].length > 0) {
            selected_items.push(order[which_group].pop());
          }
          which_group++;
          
        }
        final_set = final_set.concat(selected_items); // add the items to the end of the array
      }
    }
  }
}
```

**Reply by ghost** [reply]
Yes! Just needed to add `return final_set` at the end, but then it works perfectly. Thank you so much Josh!! Appreciate all your help :)

**Comment by egaudrain** [comment]
Hi Marita,

I think these questions are perhaps a bit too specific to be addressed here.

It's gonna take a while to read your code and understand it enough to be able to answer your first question.

For the second, I would only suggest to add in the `on_finish` callback of your block something to edit the data 'manually' before you save it. The DataCollection objects of jsPsych let you add properties to all items with the `addToAll` method ([here](https://www.jspsych.org/core_library/jspsych-data/#addtoall)), so if you manage to restrict to the trials of interest, that may be easy. One way to do this is to use the node-id, and filter using a regular expression. This is not real working code, but something that can give you an idea of how it could work:

```javascript
// Getting the last html-button-response trial
var last_hbr = jsPsych.data.get().filter({trial_type: 'html-button-response'}).last();
// We get the node id and we "parse" it
var node_id = last_hbr.values().internal_node_id.split("-");
// We remove the end portion (you may actually want to remove less... may need some tinkering)
var node_id_pattern = node_id.slice(0,-1).join("-")+"-";
// Finally we select the trials of interest and add the recall data to them
jsPsych.data.get()
    .filter({trial_type: 'html-button-response'})
    .filterCustom(function(r){ return r.internal_node_id.startsWith(node_id_pattern); })
    .addToAll(data.recall);
```

I only hope that the result of a `filterCustom()` call remains linked to the original data and that the `addToAll()` is then reflected on the `jsPsych.data` as a whole...

Hope that helps.
